core/kernel/user_ta.c

```diff
@@ -458,9 +458,23 @@ TEE_Result tee_ta_init_user_ta_session(const TEE_UUID *uuid,
 	TEE_Result res = TEE_SUCCESS;
 	struct user_ta_ctx *utc = NULL;
 
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Before updating the context list check again if a context
+	 * already exists in the list for a single instance TA.
+	 */
+	res = tee_ta_init_session_with_context(s, uuid);
```

Comment:
I don't like calling `tee_ta_init_session_with_context()` twice. How about splitting `tee_ta_init_user_ta_session()` (and `stmm_init_session()`) into a two-step operation? The first function, for instance `tee_ta_init_user_ta_session()` does everything up to the first `mutex_unlock()`, but with the mutex held by the calling function. The second function, `tee_ta_complete_user_ta_session()` is called after `tee_ta_init_user_ta_session()` has returned OK without the mutex held, instead it takes the mutex internally as needed.

Reply:
Note calling `tee_ta_init_session_with_context()` twice means that each time a session is opened on an already loaded single instance TA, we start fetching in the TA binary (early-ta or secstore-ta or nsec-fs-ta), then once done, we see that its already loaded, hence releasing the resources. It is worth it?

Reply:
I was thinking more along https://github.com/jenswi-linaro/optee_os/tree/race_fix_example
It's not even compile-tested and I think the mutex lock/unlock sequences have room for improvement.

Reply:
Thanks for the proposal. I'll update this series straight replacing some of my commits with a commit based on your proposal.